File: Data Structures & Algorithms/longest-common-subsequence/submission-1.cpp
class Solution {
public:
vector<vector<int>> dp;
    int longestCommonSubsequence(string text1, string text2) {
        dp=vector<vector<int>>(text1.size(),vector<int>(text2.size(),-1));
        return dfs(text1,text2,0,0);
    }
    int dfs(string &text1,string &text2,int i,int j){
        if(i==text1.size() || j==text2.size()) return 0;
        if(dp[i][j]!=-1) return dp[i][j];
        if(text1[i]==text2[j]) return 1+dfs(text1,text2,i+1,j+1);
        return dp[i][j]=max(dfs(text1,text2,i+1,j),dfs(text1,text2,i,j+1));
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n=len(nums)
        dp=[-1]*n
        def dfs(i):
            if dp[i]!=-1:
                return dp[i]
            
            num=1
            for j in range(i+1,n):
                if nums[i]<nums[j]:
                    num=max(num,1+dfs(j))

            dp[i]=num            
            return num

        return max(dfs(i) for i in range(n))